File: go_engine/engine/persistence.go
package engine

import (
	"database/sql"
	"encoding/json"
	"sync"
	"time"

	_ "modernc.org/sqlite"
)

var schemaStatements = []string{
	`CREATE TABLE IF NOT EXISTS workflows (
		workflow_id TEXT PRIMARY KEY,
		workflow_type TEXT NOT NULL,
		status TEXT NOT NULL,
		created_at REAL NOT NULL,
		updated_at REAL NOT NULL,
		current_run_id TEXT NOT NULL,
		run_timeout_seconds INTEGER NOT NULL
	)`,
	`CREATE TABLE IF NOT EXISTS workflow_runs (
		run_id TEXT PRIMARY KEY,
		workflow_id TEXT NOT NULL,
		workflow_type TEXT NOT NULL,
		status TEXT NOT NULL,
		state TEXT NOT NULL,
		state_data TEXT NOT NULL,
		created_at REAL NOT NULL,
		updated_at REAL NOT NULL,
		last_event_id INTEGER DEFAULT 0,
		last_heartbeat_at REAL,
		workflow_task_timeout_seconds INTEGER NOT NULL,
		run_timeout_seconds INTEGER NOT NULL
	)`,
	`CREATE TABLE IF NOT EXISTS history_events (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		run_id TEXT NOT NULL,
		event_type TEXT NOT NULL,
		timestamp REAL NOT NULL,
		attributes TEXT NOT NULL
	)`,
	`CREATE TABLE IF NOT EXISTS tasks (
		task_id TEXT PRIMARY KEY,
		run_id TEXT NOT NULL,
		task_type TEXT NOT NULL,
		queue_name TEXT NOT NULL,
		status TEXT NOT NULL,
		scheduled_at REAL NOT NULL,
		leased_until REAL,
		completed_at REAL,
		attempt INTEGER NOT NULL,
		max_attempts INTEGER NOT NULL,
		payload TEXT NOT NULL,
		last_heartbeat_at REAL,
		heartbeat_timeout_seconds INTEGER,
		timeout_at REAL,
		last_error TEXT
	)`,
	`CREATE INDEX IF NOT EXISTS idx_tasks_queue ON tasks(queue_name, status, scheduled_at)`,
	`CREATE INDEX IF NOT EXISTS idx_events_run ON history_events(run_id, id)`,
	`CREATE INDEX IF NOT EXISTS idx_runs_status ON workflow_runs(status, updated_at)`,
}

type SqlitePersistence struct {
	db  *sql.DB
	mu  sync.Mutex
}

func NewSqlitePersistence(path string) (*SqlitePersistence, error) {
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, err
	}
	p := &SqlitePersistence{db: db}
	if err := p.Initialize(); err != nil {
		return nil, err
	}
	return p, nil
}

func (p *SqlitePersistence) Initialize() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if _, err := p.db.Exec("PRAGMA journal_mode=WAL"); err != nil {
		return err
	}
	for _, stmt := range schemaStatements {
		if _, err := p.db.Exec(stmt); err != nil {
			return err
		}
	}
	return nil
}

func (p *SqlitePersistence) now() float64 {
	return float64(time.Now().UnixNano()) / 1e9
}

func (p *SqlitePersistence) CreateRun(workflowID, workflowType, runID, state string, stateData map[string]interface{}, runTimeoutSeconds int, workflowTaskTimeoutSeconds int, inputPayload map[string]interface{}) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	now := p.now()
	stateJSON, _ := json.Marshal(stateData)
	inputJSON, _ := json.Marshal(map[string]interface{}{"input": inputPayload})
	_, err := p.db.Exec(
		`INSERT INTO workflows (workflow_id, workflow_type, status, created_at, updated_at, current_run_id, run_timeout_seconds)
		 VALUES (?, ?, ?, ?, ?, ?, ?)`,
		workflowID, workflowType, "RUNNING", now, now, runID, runTimeoutSeconds,
	)
	if err != nil {
		return err
	}
	_, err = p.db.Exec(
		`INSERT INTO workflow_runs (run_id, workflow_id, workflow_type, status, state, state_data, created_at, updated_at, workflow_task_timeout_seconds, run_timeout_seconds)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		runID, workflowID, workflowType, "RUNNING", state, string(stateJSON), now, now, workflowTaskTimeoutSeconds, runTimeoutSeconds,
	)
	if err != nil {
		return err
	}
	_, err = p.db.Exec(
		`INSERT INTO history_events (run_id, event_type, timestamp, attributes)
		 VALUES (?, ?, ?, ?)`,
		runID, "WorkflowStarted", now, string(inputJSON),
	)
	return err
}

func (p *SqlitePersistence) AppendEvent(runID, eventType string, attributes map[string]interface{}) (HistoryEvent, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	now := p.now()
	attrJSON, _ := json.Marshal(attributes)
	res, err := p.db.Exec(
		`INSERT INTO history_events (run_id, event_type, timestamp, attributes)
		 VALUES (?, ?, ?, ?)`,
		runID, eventType, now, string(attrJSON),
	)
	if err != nil {
		return HistoryEvent{}, err
	}
	eventID, _ := res.LastInsertId()
	_, _ = p.db.Exec(`UPDATE workflow_runs SET last_event_id=?, updated_at=? WHERE run_id=?`, eventID, now, runID)
	return HistoryEvent{
		EventID:    eventID,
		RunID:      runID,
		EventType:  eventType,
		Timestamp:  now,
		Attributes: attributes,
	}, nil
}

func (p *SqlitePersistence) ListEvents(runID string) ([]HistoryEvent, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	rows, err := p.db.Query(`SELECT id, run_id, event_type, timestamp, attributes FROM history_events WHERE run_id=? ORDER BY id ASC`, runID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var events []HistoryEvent
	for rows.Next() {
		var id int64
		var rid, eventType string
		var ts float64
		var attrs string
		if err := rows.Scan(&id, &rid, &eventType, &ts, &attrs); err != nil {
			return nil, err
		}
		events = append(events, HistoryEvent{
			EventID:    id,
			RunID:      rid,
			EventType:  eventType,
			Timestamp:  ts,
			Attributes: decodeJSON(attrs),
		})
	}
	return events, nil
}

func (p *SqlitePersistence) GetRun(runID string) (map[string]interface{}, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	row := p.db.QueryRow(`SELECT run_id, workflow_id, workflow_type, status, state, state_data, created_at, updated_at, workflow_task_timeout_seconds, run_timeout_seconds FROM workflow_runs WHERE run_id=?`, runID)
	var rid, wid, wtype, status, state, stateData string
	var created, updated float64
	var wfTimeout, runTimeout int
	err := row.Scan(&rid, &wid, &wtype, &status, &state, &stateData, &created, &updated, &wfTimeout, &runTimeout)
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{
		"run_id":                   rid,
		"workflow_id":              wid,
		"workflow_type":            wtype,
		"status":                   status,
		"state":                    state,
		"state_data":               decodeJSON(stateData),
		"created_at":               created,
		"updated_at":               updated,
		"workflow_task_timeout_seconds": wfTimeout,
		"run_timeout_seconds":      runTimeout,
	}, nil
}

func (p *SqlitePersistence) UpdateRunState(runID, state string, stateData map[string]interface{}, status string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	now := p.now()
	stateJSON, _ := json.Marshal(stateData)
	_, err := p.db.Exec(`UPDATE workflow_runs SET state=?, state_data=?, status=?, updated_at=? WHERE run_id=?`, state, string(stateJSON), status, now, runID)
	if err != nil {
		return err
	}
	_, _ = p.db.Exec(`UPDATE workflows SET status=?, updated_at=? WHERE current_run_id=?`, status, now, runID)
	return nil
}

func (p *SqlitePersistence) InsertTask(task Task) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	payloadJSON, _ := json.Marshal(task.Payload)
	_, err := p.db.Exec(
		`INSERT INTO tasks (task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at, attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		task.TaskID, task.RunID, task.TaskType, task.QueueName, task.Status, task.ScheduledAt, task.LeasedUntil, task.CompletedAt, task.Attempt, task.MaxAttempts, string(payloadJSON), task.LastHeartbeatAt, task.HeartbeatTimeoutSeconds, task.TimeoutAt, task.LastError,
	)
	return err
}

func (p *SqlitePersistence) UpdateTaskStatus(taskID, status string, completedAt *float64, lastError *string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	_, err := p.db.Exec(`UPDATE tasks SET status=?, completed_at=?, last_error=? WHERE task_id=?`, status, completedAt, lastError, taskID)
	return err
}

func (p *SqlitePersistence) UpdateTaskSchedule(taskID string, scheduledAt float64) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	_, err := p.db.Exec(`UPDATE tasks SET status='PENDING', scheduled_at=?, leased_until=NULL WHERE task_id=?`, scheduledAt, taskID)
	return err
}

func (p *SqlitePersistence) UpdateTaskHeartbeat(taskID string, heartbeatAt float64) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	_, err := p.db.Exec(`UPDATE tasks SET last_heartbeat_at=? WHERE task_id=?`, heartbeatAt, taskID)
	return err
}

func (p *SqlitePersistence) GetTask(taskID string) (*Task, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	row := p.db.QueryRow(`SELECT task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at, attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error FROM tasks WHERE task_id=?`, taskID)
	var task Task
	var payload string
	var leasedUntil, completedAt, lastHeartbeatAt, timeoutAt sql.NullFloat64
	var heartbeatTimeout sql.NullInt64
	var lastError sql.NullString
	if err := row.Scan(&task.TaskID, &task.RunID, &task.TaskType, &task.QueueName, &task.Status, &task.ScheduledAt, &leasedUntil, &completedAt, &task.Attempt, &task.MaxAttempts, &payload, &lastHeartbeatAt, &heartbeatTimeout, &timeoutAt, &lastError); err != nil {
		return nil, err
	}
	task.Payload = decodeJSON(payload)
	if leasedUntil.Valid {
		task.LeasedUntil = &leasedUntil.Float64
	}
	if completedAt.Valid {
		task.CompletedAt = &completedAt.Float64
	}
	if lastHeartbeatAt.Valid {
		task.LastHeartbeatAt = &lastHeartbeatAt.Float64
	}
	if heartbeatTimeout.Valid {
		val := int(heartbeatTimeout.Int64)
		task.HeartbeatTimeoutSeconds = &val
	}
	if timeoutAt.Valid {
		task.TimeoutAt = &timeoutAt.Float64
	}
	if lastError.Valid {
		task.LastError = &lastError.String
	}
	return &task, nil
}

func (p *SqlitePersistence) LeaseTasks(queueName string, limit int, leaseSeconds int) ([]Task, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	now := p.now()
	tx, err := p.db.Begin()
	if err != nil {
		return nil, err
	}
	rows, err := tx.Query(
		`SELECT task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at, attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error
		 FROM tasks
		 WHERE queue_name=?
		   AND ((status='PENDING' AND scheduled_at<=?) OR (status='LEASED' AND leased_until<=?))
		   AND (timeout_at IS NULL OR timeout_at>=?)
		 ORDER BY scheduled_at ASC
		 LIMIT ?`,
		queueName, now, now, now, limit,
	)
	if err != nil {
		_ = tx.Rollback()
		return nil, err
	}
	defer rows.Close()
	var leased []Task
	for rows.Next() {
		var task Task
		var payload string
		var leasedUntil, completedAt, lastHeartbeatAt, timeoutAt sql.NullFloat64
		var heartbeatTimeout sql.NullInt64
		var lastError sql.NullString
		if err := rows.Scan(&task.TaskID, &task.RunID, &task.TaskType, &task.QueueName, &task.Status, &task.ScheduledAt, &leasedUntil, &completedAt, &task.Attempt, &task.MaxAttempts, &payload, &lastHeartbeatAt, &heartbeatTimeout, &timeoutAt, &lastError); err != nil {
			_ = tx.Rollback()
			return nil, err
		}
		task.Payload = decodeJSON(payload)
		if task.Attempt >= task.MaxAttempts {
			err = p.updateTaskStatusTx(tx, task.TaskID, "FAILED", nil, strPtr("max_attempts_exceeded"))
			if err != nil {
				_ = tx.Rollback()
				return nil, err
			}
			continue
		}
		attempt := task.Attempt + 1
		leasedUntilVal := now + float64(leaseSeconds)
		_, err := tx.Exec(`UPDATE tasks SET status='LEASED', leased_until=?, attempt=?, last_heartbeat_at=? WHERE task_id=?`, leasedUntilVal, attempt, now, task.TaskID)
		if err != nil {
			_ = tx.Rollback()
			return nil, err
		}
		task.Status = "LEASED"
		task.Attempt = attempt
		task.LeasedUntil = &leasedUntilVal
		task.LastHeartbeatAt = &now
		leased = append(leased, task)
	}
	if err := tx.Commit(); err != nil {
		return nil, err
	}
	return leased, nil
}

func (p *SqlitePersistence) updateTaskStatusTx(tx *sql.Tx, taskID, status string, completedAt *float64, lastError *string) error {
	_, err := tx.Exec(`UPDATE tasks SET status=?, completed_at=?, last_error=? WHERE task_id=?`, status, completedAt, lastError, taskID)
	return err
}

func (p *SqlitePersistence) FindHeartbeatExpiredTasks(now float64) ([]Task, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	rows, err := p.db.Query(
		`SELECT task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at, attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error
		 FROM tasks
		 WHERE status='LEASED'
		   AND heartbeat_timeout_seconds IS NOT NULL
		   AND last_heartbeat_at IS NOT NULL
		   AND last_heartbeat_at + heartbeat_timeout_seconds <= ?`,
		now,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Task
	for rows.Next() {
		task, err := scanTask(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, task)
	}
	return out, nil
}

func (p *SqlitePersistence) FindTimeoutExpiredTasks(now float64) ([]Task, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	rows, err := p.db.Query(
		`SELECT task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at, attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error
		 FROM tasks
		 WHERE status='LEASED'
		   AND timeout_at IS NOT NULL
		   AND timeout_at <= ?`,
		now,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []Task
	for rows.Next() {
		task, err := scanTask(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, task)
	}
	return out, nil
}

func (p *SqlitePersistence) FindRunTimeouts(now float64) ([]map[string]interface{}, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	rows, err := p.db.Query(
		`SELECT run_id, workflow_id, workflow_type, state, state_data FROM workflow_runs WHERE status='RUNNING' AND created_at + run_timeout_seconds <= ?`,
		now,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []map[string]interface{}
	for rows.Next() {
		var runID, workflowID, workflowType, state, stateData string
		if err := rows.Scan(&runID, &workflowID, &workflowType, &state, &stateData); err != nil {
			return nil, err
		}
		out = append(out, map[string]interface{}{
			"run_id":       runID,
			"workflow_id":  workflowID,
			"workflow_type": workflowType,
			"state":        state,
			"state_data":   decodeJSON(stateData),
		})
	}
	return out, nil
}

func scanTask(rows *sql.Rows) (Task, error) {
	var task Task
	var payload string
	var leasedUntil, completedAt, lastHeartbeatAt, timeoutAt sql.NullFloat64
	var heartbeatTimeout sql.NullInt64
	var lastError sql.NullString
	err := rows.Scan(&task.TaskID, &task.RunID, &task.TaskType, &task.QueueName, &task.Status, &task.ScheduledAt, &leasedUntil, &completedAt, &task.Attempt, &task.MaxAttempts, &payload, &lastHeartbeatAt, &heartbeatTimeout, &timeoutAt, &lastError)
	if err != nil {
		return task, err
	}
	task.Payload = decodeJSON(payload)
	if leasedUntil.Valid {
		task.LeasedUntil = &leasedUntil.Float64
	}
	if completedAt.Valid {
		task.CompletedAt = &completedAt.Float64
	}
	if lastHeartbeatAt.Valid {
		task.LastHeartbeatAt = &lastHeartbeatAt.Float64
	}
	if heartbeatTimeout.Valid {
		val := int(heartbeatTimeout.Int64)
		task.HeartbeatTimeoutSeconds = &val
	}
	if timeoutAt.Valid {
		task.TimeoutAt = &timeoutAt.Float64
	}
	if lastError.Valid {
		task.LastError = &lastError.String
	}
	return task, nil
}

func strPtr(value string) *string {
	return &value
}

File: python_engine/workflow_engine/persistence.py
"""
SQLite persistence layer with explicit schemas.
"""

from __future__ import annotations

import json
import sqlite3
import threading
import time
from typing import Any, Dict, Iterable, List, Optional

from .models import HistoryEvent, Task


SCHEMA_STATEMENTS = [
    """
    CREATE TABLE IF NOT EXISTS workflows (
        workflow_id TEXT PRIMARY KEY,
        workflow_type TEXT NOT NULL,
        status TEXT NOT NULL,
        created_at REAL NOT NULL,
        updated_at REAL NOT NULL,
        current_run_id TEXT NOT NULL,
        run_timeout_seconds INTEGER NOT NULL
    )
    """,
    """
    CREATE TABLE IF NOT EXISTS workflow_runs (
        run_id TEXT PRIMARY KEY,
        workflow_id TEXT NOT NULL,
        workflow_type TEXT NOT NULL,
        status TEXT NOT NULL,
        state TEXT NOT NULL,
        state_data TEXT NOT NULL,
        created_at REAL NOT NULL,
        updated_at REAL NOT NULL,
        last_event_id INTEGER DEFAULT 0,
        last_heartbeat_at REAL,
        workflow_task_timeout_seconds INTEGER NOT NULL,
        run_timeout_seconds INTEGER NOT NULL
    )
    """,
    """
    CREATE TABLE IF NOT EXISTS history_events (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        run_id TEXT NOT NULL,
        event_type TEXT NOT NULL,
        timestamp REAL NOT NULL,
        attributes TEXT NOT NULL
    )
    """,
    """
    CREATE TABLE IF NOT EXISTS tasks (
        task_id TEXT PRIMARY KEY,
        run_id TEXT NOT NULL,
        task_type TEXT NOT NULL,
        queue_name TEXT NOT NULL,
        status TEXT NOT NULL,
        scheduled_at REAL NOT NULL,
        leased_until REAL,
        completed_at REAL,
        attempt INTEGER NOT NULL,
        max_attempts INTEGER NOT NULL,
        payload TEXT NOT NULL,
        last_heartbeat_at REAL,
        heartbeat_timeout_seconds INTEGER,
        timeout_at REAL,
        last_error TEXT
    )
    """,
    "CREATE INDEX IF NOT EXISTS idx_tasks_queue ON tasks(queue_name, status, scheduled_at)",
    "CREATE INDEX IF NOT EXISTS idx_events_run ON history_events(run_id, id)",
    "CREATE INDEX IF NOT EXISTS idx_runs_status ON workflow_runs(status, updated_at)",
]


class SqlitePersistence:
    def __init__(self, db_path: str) -> None:
        self._db_path = db_path
        self._conn = sqlite3.connect(self._db_path, check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        self._lock = threading.Lock()
        self.initialize()

    def initialize(self) -> None:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute("PRAGMA journal_mode=WAL")
            for statement in SCHEMA_STATEMENTS:
                cur.execute(statement)
            self._conn.commit()

    def create_run(
        self,
        workflow_id: str,
        workflow_type: str,
        run_id: str,
        state: str,
        state_data: Dict[str, Any],
        run_timeout_seconds: int,
        workflow_task_timeout_seconds: int,
        input_payload: Dict[str, Any],
    ) -> None:
        now = time.time()
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                INSERT INTO workflows (workflow_id, workflow_type, status, created_at, updated_at, current_run_id, run_timeout_seconds)
                VALUES (?, ?, ?, ?, ?, ?, ?)
                """,
                (workflow_id, workflow_type, "RUNNING", now, now, run_id, run_timeout_seconds),
            )
            cur.execute(
                """
                INSERT INTO workflow_runs (run_id, workflow_id, workflow_type, status, state, state_data, created_at, updated_at, workflow_task_timeout_seconds, run_timeout_seconds)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    run_id,
                    workflow_id,
                    workflow_type,
                    "RUNNING",
                    state,
                    json.dumps(state_data),
                    now,
                    now,
                    workflow_task_timeout_seconds,
                    run_timeout_seconds,
                ),
            )
            cur.execute(
                """
                INSERT INTO history_events (run_id, event_type, timestamp, attributes)
                VALUES (?, ?, ?, ?)
                """,
                (run_id, "WorkflowStarted", now, json.dumps({"input": input_payload})),
            )
            self._conn.commit()

    def append_event(self, run_id: str, event_type: str, attributes: Dict[str, Any]) -> HistoryEvent:
        now = time.time()
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                INSERT INTO history_events (run_id, event_type, timestamp, attributes)
                VALUES (?, ?, ?, ?)
                """,
                (run_id, event_type, now, json.dumps(attributes)),
            )
            event_id = cur.lastrowid
            cur.execute(
                "UPDATE workflow_runs SET last_event_id=?, updated_at=? WHERE run_id=?",
                (event_id, now, run_id),
            )
            self._conn.commit()
        return HistoryEvent(event_id=event_id, run_id=run_id, event_type=event_type, timestamp=now, attributes=attributes)

    def list_events(self, run_id: str) -> List[HistoryEvent]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                "SELECT * FROM history_events WHERE run_id=? ORDER BY id ASC",
                (run_id,),
            )
            rows = cur.fetchall()
        return [HistoryEvent.from_row(row) for row in rows]

    def get_run(self, run_id: str) -> Optional[Dict[str, Any]]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute("SELECT * FROM workflow_runs WHERE run_id=?", (run_id,))
            row = cur.fetchone()
        if not row:
            return None
        return {
            "run_id": row["run_id"],
            "workflow_id": row["workflow_id"],
            "workflow_type": row["workflow_type"],
            "status": row["status"],
            "state": row["state"],
            "state_data": json.loads(row["state_data"] or "{}"),
            "created_at": row["created_at"],
            "updated_at": row["updated_at"],
            "workflow_task_timeout_seconds": row["workflow_task_timeout_seconds"],
            "run_timeout_seconds": row["run_timeout_seconds"],
        }

    def update_run_state(self, run_id: str, state: str, state_data: Dict[str, Any], status: str) -> None:
        now = time.time()
        with self._lock:
            self._conn.execute(
                """
                UPDATE workflow_runs
                SET state=?, state_data=?, status=?, updated_at=?
                WHERE run_id=?
                """,
                (state, json.dumps(state_data), status, now, run_id),
            )
            self._conn.execute(
                "UPDATE workflows SET status=?, updated_at=? WHERE current_run_id=?",
                (status, now, run_id),
            )
            self._conn.commit()

    def insert_task(self, task: Task) -> None:
        with self._lock:
            self._conn.execute(
                """
                INSERT INTO tasks (task_id, run_id, task_type, queue_name, status, scheduled_at, leased_until, completed_at,
                                   attempt, max_attempts, payload, last_heartbeat_at, heartbeat_timeout_seconds, timeout_at, last_error)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    task.task_id,
                    task.run_id,
                    task.task_type,
                    task.queue_name,
                    task.status,
                    task.scheduled_at,
                    task.leased_until,
                    task.completed_at,
                    task.attempt,
                    task.max_attempts,
                    json.dumps(task.payload),
                    task.last_heartbeat_at,
                    task.heartbeat_timeout_seconds,
                    task.timeout_at,
                    task.last_error,
                ),
            )
            self._conn.commit()

    def update_task_status(self, task_id: str, status: str, completed_at: Optional[float] = None, last_error: Optional[str] = None) -> None:
        with self._lock:
            self._conn.execute(
                """
                UPDATE tasks SET status=?, completed_at=?, last_error=?
                WHERE task_id=?
                """,
                (status, completed_at, last_error, task_id),
            )
            self._conn.commit()

    def update_task_lease(self, task_id: str, leased_until: float, attempt: int) -> None:
        with self._lock:
            self._conn.execute(
                """
                UPDATE tasks SET status='LEASED', leased_until=?, attempt=?, last_heartbeat_at=?
                WHERE task_id=?
                """,
                (leased_until, attempt, time.time(), task_id),
            )
            self._conn.commit()

    def update_task_schedule(self, task_id: str, scheduled_at: float) -> None:
        with self._lock:
            self._conn.execute(
                """
                UPDATE tasks SET status='PENDING', scheduled_at=?, leased_until=NULL
                WHERE task_id=?
                """,
                (scheduled_at, task_id),
            )
            self._conn.commit()

    def update_task_heartbeat(self, task_id: str, heartbeat_at: float) -> None:
        with self._lock:
            self._conn.execute(
                """
                UPDATE tasks SET last_heartbeat_at=?
                WHERE task_id=?
                """,
                (heartbeat_at, task_id),
            )
            self._conn.commit()

    def get_task(self, task_id: str) -> Optional[Task]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute("SELECT * FROM tasks WHERE task_id=?", (task_id,))
            row = cur.fetchone()
        if not row:
            return None
        return Task.from_row(row)

    def list_tasks(self, task_ids: Iterable[str]) -> List[Task]:
        ids = list(task_ids)
        if not ids:
            return []
        placeholders = ",".join("?" for _ in ids)
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(f"SELECT * FROM tasks WHERE task_id IN ({placeholders})", ids)
            rows = cur.fetchall()
        return [Task.from_row(row) for row in rows]

    def find_candidate_tasks(self, queue_name: str, now: float, limit: int) -> List[Task]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT * FROM tasks
                WHERE queue_name=?
                AND (
                    (status='PENDING' AND scheduled_at<=?)
                    OR (status='LEASED' AND leased_until<=?)
                )
                AND (timeout_at IS NULL OR timeout_at>=?)
                ORDER BY scheduled_at ASC
                LIMIT ?
                """,
                (queue_name, now, now, now, limit),
            )
            rows = cur.fetchall()
        return [Task.from_row(row) for row in rows]

    def lease_tasks(self, queue_name: str, now: float, limit: int, lease_seconds: int) -> List[Task]:
        leased: List[Task] = []
        with self._lock:
            cur = self._conn.cursor()
            cur.execute("BEGIN IMMEDIATE")
            cur.execute(
                """
                SELECT * FROM tasks
                WHERE queue_name=?
                AND (
                    (status='PENDING' AND scheduled_at<=?)
                    OR (status='LEASED' AND leased_until<=?)
                )
                AND (timeout_at IS NULL OR timeout_at>=?)
                ORDER BY scheduled_at ASC
                LIMIT ?
                """,
                (queue_name, now, now, now, limit),
            )
            rows = cur.fetchall()
            for row in rows:
                task = Task.from_row(row)
                if task.attempt >= task.max_attempts:
                    cur.execute(
                        "UPDATE tasks SET status='FAILED', last_error=? WHERE task_id=?",
                        ("max_attempts_exceeded", task.task_id),
                    )
                    continue
                attempt = task.attempt + 1
                leased_until = now + lease_seconds
                cur.execute(
                    """
                    UPDATE tasks
                    SET status='LEASED', leased_until=?, attempt=?, last_heartbeat_at=?
                    WHERE task_id=?
                    """,
                    (leased_until, attempt, now, task.task_id),
                )
                task.status = "LEASED"
                task.leased_until = leased_until
                task.attempt = attempt
                task.last_heartbeat_at = now
                leased.append(task)
            self._conn.commit()
        return leased

    def find_heartbeat_expired_tasks(self, now: float) -> List[Task]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT * FROM tasks
                WHERE status='LEASED'
                AND heartbeat_timeout_seconds IS NOT NULL
                AND last_heartbeat_at IS NOT NULL
                AND last_heartbeat_at + heartbeat_timeout_seconds <= ?
                """,
                (now,),
            )
            rows = cur.fetchall()
        return [Task.from_row(row) for row in rows]

    def find_timeout_expired_tasks(self, now: float) -> List[Task]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT * FROM tasks
                WHERE status='LEASED'
                AND timeout_at IS NOT NULL
                AND timeout_at <= ?
                """,
                (now,),
            )
            rows = cur.fetchall()
        return [Task.from_row(row) for row in rows]

    def find_run_timeouts(self, now: float) -> List[Dict[str, Any]]:
        with self._lock:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT * FROM workflow_runs
                WHERE status='RUNNING'
                AND created_at + run_timeout_seconds <= ?
                """,
                (now,),
            )
            rows = cur.fetchall()
        return [
            {
                "run_id": row["run_id"],
                "workflow_id": row["workflow_id"],
                "workflow_type": row["workflow_type"],
                "state": row["state"],
                "state_data": json.loads(row["state_data"] or "{}"),
            }
            for row in rows
        ]
